File: structural_patterns/composite/go/file_system/folder.go
package filesystem

import "fmt"

type Folder struct {
	components []component
	name       string
}

func NewFolder(name string) *Folder {
	return &Folder{name: name}
}

func (f *Folder) Search(key string) {
	fmt.Printf("Searching '%s' in folder '%s':\n", key, f.name)

	for _, component := range f.components {
		component.Search(key)
	}
}

func (f *Folder) Add(comp component) {
	f.components = append(f.components, comp)
}

File: structural_patterns/composite/cpp/folder.cpp
#include "folder.h"

void Folder::search(std::string key) {
    std::cout << "Searching '" << key << "' in folder '" << name << "':\n";

    for (auto elem : components) {
		elem->search(key);
    }
}

bool Folder::add(std::initializer_list<PtrComponent> list)
{
	for(auto elem : list)
		components.push_back(elem);
	
	return true;
}
